pkg/cloud/impl_registry.go

```diff
@@ -380,3 +472,91 @@ type ReadWriterInterceptor interface {
 	Reader(context.Context, ExternalStorage, ioctx.ReadCloserCtx) ioctx.ReadCloserCtx
 	Writer(context.Context, ExternalStorage, io.WriteCloser) io.WriteCloser
 }
+
+// makeExternalStorage creates an ExternalStorage from the given config.
+func makeExternalStorage[T any](
+	ctx context.Context,
+	dest cloudpb.ExternalStorage,
+	conf base.ExternalIODirConfig,
+	limiters Limiters,
+	metrics metric.Struct,
+	args T,
+	getImpl func(cloudpb.ExternalStorageProvider) (func(context.Context, T, cloudpb.ExternalStorage) (ExternalStorage, error), bool),
+	opts ...ExternalStorageOption,
+) (ExternalStorage, error) {
+	var cloudMetrics *Metrics
+	var ok bool
+	if cloudMetrics, ok = metrics.(*Metrics); !ok {
+		return nil, errors.Newf("invalid metrics type: %T", metrics)
+	}
+	if conf.DisableOutbound && dest.Provider != cloudpb.ExternalStorageProvider_userfile {
+		return nil, errors.New("external network access is disabled")
+	}
+	options := ExternalStorageOptions{}
+	for _, o := range opts {
+		o(&options)
+	}
+	if fn, ok := getImpl(dest.Provider); ok {
+		e, err := fn(ctx, args, dest)
+		if err != nil {
+			return nil, err
+		}
+
+		// We do not wrap the ExternalStorage for the `external` provider. An
+		// external connection object represents an underlying external resource
+		// that will have its own `esWrapper`.
+		if dest.Provider == cloudpb.ExternalStorageProvider_external {
+			return e, nil
+		}
+
+		return &esWrapper{
+			ExternalStorage: e,
+			lim:             limiters[dest.Provider],
+			ioRecorder:      options.ioAccountingInterceptor,
+			metricsRecorder: newMetricsReadWriter(cloudMetrics),
+		}, nil
+	}
+
+	return nil, errors.Errorf("unsupported external destination type: %s", dest.Provider.String())
+}
+
+func HijackProviderForTesting(
```

Comment:
nit: s/Hijack/Replace/r :D

Perhaps also name the return func `cleanup` or provide a sentence in the docstring on what this function does

Reply:
or `Intercept`

Reply:
Went with Replace.

Reply:
Sorry, I know you _just_ replaced all these names, but I believe our convention is to prefix with `Testing` e.g. `TestingReplaceProvider`